# CNV-P: A machine-learning framework for filtering copy number variations
CNV-PG is a novel and postâ€“processing approach for CNV filtering.

## Prerequisites:
[python3](https://www.python.org/)
[sklearn](https://pypi.org/project/sklearn/)
[matplotlib](https://pypi.org/project/matplotlib/)
[pysam](https://pypi.org/project/pysam/)
[pandas](https://pypi.org/project/pandas/)
[numpy](https://pypi.org/project/numpy/)

## Install by conda
conda install python=3.7.0
conda install -c anaconda scikit-learn=0.21
conda install -c conda-forge matplotlib
conda install -c bioconda pysam
conda install -c anaconda pandas
conda install -c anaconda numpy

## Getting started
### 1. CNV Predicting
run the "python  script/CNV-P_predict_main.py -h" to see the USAGE;
```
usage: CNV-P_predict_main.py [-h] [-m model] -b bamfile -s CNVcaller -bed
                             BEDfile -bas basfile [-sam Samplename]
                             [-o outdir]
optional arguments:
  -h, --help            show this help message and exit
  -m model, --model model
                        which model you want to used
  -b bamfile, --bam bamfile
                        file provides features
  -s CNVcaller, --soft CNVcaller
                        which CNV caller you used
  -bed bedfile, --CNV_bed bedfile
                        the format of input file
  -bas basfile, --basfile basfile
                        file that provide mean insert-size and sequencing
                        depth
  -sam Samplename, --Samplename Samplename
                        Samplename that use as prefix of result
  -o outdir, --output outdir
                        the directory into which output files will go
```
##### 1.1 input parameters
**model**: 
**CNVcaller**:  
**bamfile**: BAM file should generated by a read aligner that supports partial read alignments, such as BWA-MEM
**bedfile**: This file should be 5 Columns: chromsome, start, end, size of CNV, type of CNV (DUP:1,DEL:0) (for example: Test_data/HG002.Lumpy.fil.mer.bed)
**basfile**: this file should be 6 columns: 
##### 1.2 output
**samplename.feature.txt**: Extracted feature matrix.
**samplename.pre.prop.txt**: The prediction result and probability score. 

##### 1.3 running example
```
python  script/CNV-P_predict_main.py  -m RF -b Test_data/HG002.test.bam -s Lumpy -bed Test_data/HG002.Lumpy.fil.mer.bed -bas Test_data/HG002.bam.bas -sam HG002 -o Test_data/out/
```

### 2. training for other CNV callers
run the "python  script/CNV-P_predict_main.py -h" to see the USAGE;
```
usage: CNV-P_training_main.py [-h] [-m model] -b bamfile -s CNVcaller -bed
                              BEDfile -bas basfile -lab labelfile [-o outdir]
optional arguments:
  -h, --help            show this help message and exit
  -m model, --model model
                        which model you want to used
  -b bamfile, --bam bamfile
                        file provides features
  -s CNVcaller, --soft CNVcaller
                        which CNV caller you used
  -bed bedfile, --CNV_bed bedfile
                        the format of input file
  -bas basfile, --basfile basfile
                        file that provide mean insert-size and sequencing
                        depth
  -lab labelfile, --labelfile labelfile
                        file that provide CNV label, true CNVs labeled as
                        1,false CNVs labeled as 0, The order should
                        corresponds to CNV_bed file(-bed/--CNV_bed) one to one
  -o outdir, --output outdir
                        the directory into which output files will go
```
##### 2.1 input parameters


##### 2.2 outputs:

##### 2.3 running example
```

```

Please help us improve CNV-P by reporting bugs or ideas on how to make things better.

